### PR TITLE
Prevent Lottie from running in the background after navigating to other pages

### PR DESCRIPTION
**Pull request by QichenZhu**
### Fixed Issues

$ https://github.com/Expensify/App/issues/47273
PROPOSAL: https://github.com/Expensify/App/issues/47273#issuecomment-2294733914

### Tests

1. Go to Settings - Preferences.
2. Navigate to Chat.
3. Navigate back to the Settings tab.
4. Click on Priority Mode, Language, or Theme.

**Expected result:** The animation in the Settings tab should work, and the right-side modal should open.

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Go to Settings - Preferences.
2. Navigate to Chat.
3. Navigate back to the Settings tab.
4. Click on Priority Mode, Language, or Theme.

**Expected result:** The animation in the Settings tab should work, and the right-side modal should open.

### QA Steps

1. Go to Settings - Preferences.
2. Navigate to Chat.
3. Navigate back to the Settings tab.
4. Click on Priority Mode, Language, or Theme.

**Expected result:** The animation in the Settings tab should work, and the right-side modal should open.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/dc7b432f-f7cc-4c68-8dd8-9d79b7591693

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/937dfd2c-cd36-4a7e-8461-1806922e2ac5

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c78f87c2-6acf-421c-a1ed-cba6ce5f7846

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/10d526b6-a67a-49ad-96ea-b30cc02b08cf

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/7c4371a4-b79e-442a-8929-3be4ec7cba7b

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/32d74405-d71c-4093-8675-92bb9bc8c300

</details>

### CODE REVIEW

File: src/components/Lottie/index.tsx [hunk 1]
```diff
@@ -27,13 +32,39 @@ function Lottie({source, webStyle, ...props}: Props, ref: ForwardedRef<LottieVie
...
+        const unsubscribeNavigationFocus = navigator.addListener('focus', () => {
+            setIsHidden(false);
+        });
+        return unsubscribeNavigationFocus;
+    }, [navigationContainerRef, navigator]);
+
+    // Prevent the animation from running in the background after navigating to other pages.
+    // See https://github.com/Expensify/App/issues/47273
+    useEffect(() => {
+        if (!navigationContainerRef || !navigator) {
```
Comment (suneox) [inline src/components/Lottie/index.tsx:48]: We should check navigationRef.isReady() before access getState

**Comment by QichenZhu** [comment]
@suneox It seems that the current approach has fixed the regression, so using `isReady()` might be redundant.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
mac-web



</details>

<details>
<summary>Android: mWeb Chrome</summary>
mac-web



</details>

<details>
<summary>iOS: Native</summary>
mac-web



</details>

<details>
<summary>iOS: mWeb Safari</summary>
mac-web



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>
mac-web



</details>

<details>
<summary>MacOS: Desktop</summary>
mac-web



</details>

**Comment by suneox** [comment]
> @suneox It seems that the current approach has fixed the regression, so using `isReady()` might be redundant.

TY, I'll start checking now

File: src/components/Lottie/index.tsx [hunk 2]
```diff
@@ -30,11 +33,40 @@ function Lottie({source, webStyle, ...props}: Props, ref: ForwardedRef<LottieVie
...
+    }, [browser, navigationContainerRef, navigator]);
+
+    useEffect(() => {
+        if (!browser || !navigationContainerRef || !navigator) {
+            return;
+        }
+        const unsubscribeNavigationBlur = navigator.addListener('blur', () => {
+            const state = navigationContainerRef.getRootState();
+            const targetRouteName = state?.routes?.[state?.index ?? 0]?.name;
+            if (targetRouteName !== NAVIGATORS.RIGHT_MODAL_NAVIGATOR) {
```
Comment (suneox) [inline src/components/Lottie/index.tsx:58]:
```suggestion
            if (![NAVIGATORS.RIGHT_MODAL_NAVIGATOR, NAVIGATORS.LEFT_MODAL_NAVIGATOR].includes(targetRouteName)) {
```

File: src/components/Lottie/index.tsx [hunk 3]
```diff
@@ -30,11 +33,40 @@ function Lottie({source, webStyle, ...props}: Props, ref: ForwardedRef<LottieVie
...
     const aspectRatioStyle = styles.aspectRatioLottie(source);
 
-    // If the image fails to load, app is in background state, animation file isn't ready, or the splash screen isn't hidden yet,
+    const browser = Browser.getBrowser();
+    const [hasNavigatedAway, setHasNavigatedAway] = React.useState(false);
+    const navigationContainerRef = useContext(NavigationContainerRefContext);
+    const navigator = useContext(NavigationContext);
+
+    useEffect(() => {
+        if (!browser || !navigationContainerRef || !navigator) {
```
Comment (suneox) [inline src/components/Lottie/index.tsx:42]: Do we need to check browser?

**Comment by suneox** [comment]
@QichenZhu The change looks good. We can update the status based on the suggestions above, and I have a question related to using the browser.

File: src/components/Lottie/index.tsx (hunk 2 above)
Comment (QichenZhu) [reply src/components/Lottie/index.tsx]: That will be better. Thanks for pointing it out!

File: src/components/Lottie/index.tsx (hunk 3 above)
Comment (QichenZhu) [reply src/components/Lottie/index.tsx:42]: Because the issue only occurs in browsers, it seems unnecessary to apply this fix to native platforms.

Comment (suneox) [reply src/components/Lottie/index.tsx:42]: OK, let change the PR status

**Comment by dangrous** [comment]
No worries - looks like this should be reviewed by @aldo-expensify ?

**Comment by suneox** [comment]
Hi @dangrous I’ll be working on this PR

**Comment by dangrous** [comment]
ah @QichenZhu it looks like you just linked the PR like `$ #47273` - next time be sure to put the full link in the markdown, like `$ https://github.com/Expensify/App/issues/47273` - that will let the bot correctly assigned reviewers!

**Comment by suneox** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/user-attachments/assets/ccd3d438-c770-432d-ac41-fd0fe84e9b7a



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/user-attachments/assets/1aae0d36-c5ea-4925-b297-df1532144353



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/user-attachments/assets/6a89f56e-9eec-468e-889a-4f8589436146



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/user-attachments/assets/0208abd5-e7f1-47a9-8fb5-ee36c7ef0155



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/user-attachments/assets/74d6930b-9fa3-4751-8d02-69d113b956f6



</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/user-attachments/assets/9d666da1-3616-4a22-968a-db4f15ad836d



</details>

**Review by suneox — APPROVED** [review]
LGTM. I’ve verified this issue along with the regression issue.

**Review by aldo-expensify — APPROVED** [review]

**Pull request merged**